### ISSUE TITLE
Running jupyter-lab a 3rd time fails

### ISSUE THREAD

**Issue by djacu** [tweag/jupyterWith#233]
**Describe the bug**
JupyterLab will not start a 3rd time. The first time you run `./result/bin/jupyter-lab`, IPython complains that it has no writeable path as documented in https://github.com/tweag/jupyterWith/issues/230#issuecomment-1194861045. The second time, everything appears to work normally. The third time an error is thrown and JupyterLab will not start.

**To Reproduce**
Build the project. Run jupyter-lab 3 times while opening the web app each time you run it.

Build
```shell
nix build .#example_jupyterlab
```

Run JupyterLab
```shell
./result/bin/jupyter-lab
```

Open Web App. Shutdown server. Repeat from Run JupyterLab 2 more times and it should fail the 2nd time.

**Expected behavior**
JupyterLab should not fail.

**Environment**
 - NixOS-22.11
 - commit: d1a9b1eaac101e2c01d1c10846da10b0b5e790eb

**Additional context**
Shell output from running JupyterLab the 3rd time.

```shell
$ ./result/bin/jupyter-lab
Traceback (most recent call last):
  File "/nix/store/4xyz6l1rc06y0bsgc4q86yqq9iwr3v1i-python3.10-jupyterlab-3.4.3/bin/.jupyter-lab-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/jupyter_server/extension/application.py", line 584, in launch_instance
    serverapp = cls.initialize_server(argv=args)
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/jupyter_server/extension/application.py", line 554, in initialize_server
    serverapp.initialize(
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/traitlets/config/application.py", line 110, in inner
    return method(app, *args, **kwargs)
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/jupyter_server/serverapp.py", line 2435, in initialize
    super().initialize(argv=argv)
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/traitlets/config/application.py", line 110, in inner
    return method(app, *args, **kwargs)
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/jupyter_core/application.py", line 244, in initialize
    self.migrate_config()
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/jupyter_core/application.py", line 174, in migrate_config
    migrate()
  File "/nix/store/cl3hbbkpahmpq7v3lmk1ps15mp01wg2r-python3-3.10.4-env/lib/python3.10/site-packages/jupyter_core/migrate.py", line 241, in migrate
    with open(os.path.join(env["jupyter_config"], "migrated"), "w", encoding="utf-8") as f:
OSError: [Errno 30] Read-only file system: '/nix/store/mkgm8n6yp0645s8fpsfjsap3nwlnxwyq-jupyter-dir/config/migrated'
```

**Comment by djacu** [comment]
Most applications in the Jupyter ecosystem inherit from [JupyterApp](https://github.com/jupyter/jupyter_core/blob/550519504c19b8d7529dad7bdc379e72df14b0b8/jupyter_core/application.py#L65). They will define an `initialize` method and call `super().initialize`, which will eventually lead back to `JupyterApp`. In `JupyterApp` it calls [migrate_config](https://github.com/jupyter/jupyter_core/blob/550519504c19b8d7529dad7bdc379e72df14b0b8/jupyter_core/application.py#L244) which imports a `migrate` function and runs it. `migrate` looks for IPython configurations and attempts to migrate them to Jupyter. This issue is when it tries to [write the new config](https://github.com/jupyter/jupyter_core/blob/550519504c19b8d7529dad7bdc379e72df14b0b8/jupyter_core/migrate.py#L241). The config path is in the nix store which we put there so it can be managed decoratively.

There are two checks in the `migrate_config` [function](https://github.com/jupyter/jupyter_core/blob/550519504c19b8d7529dad7bdc379e72df14b0b8/jupyter_core/application.py#L162) that gives us an escape to fix this.

1. It checks if a `migrated` folder already exists. If it does, it will exit and do nothing.
2. If there is no IPythond directory, it will exit.

Both appear possible but it might be better to a recent commit where we set the `IPYTHONDIR` env variable back to an non-writable path.

**Comment by djacu** [comment]
The reason this breaks on the 3rd time is as follows:

1. The 1st time, `.jupyter` does not exist when IPython tries to write to the `.jupyter/.ipython` folder. So it uses a temporary directory. Jupyter will create the `.jupyter` folder but only after IPython tries to write and it doesn't try a second time.
2. The 2nd time `.jupyter` folder does exist so IPython will put a folder there (e.g. `.jupyter/.ipython`). 
3. The 3rd time, JupyterServer sees the IPython directory, does not see a migrated directory, and then attempts to migrate the IPython configs. However, we have set the Jupyter configs to be in the nix store, so Jupyter cannot write there and throws an exception.

**Comment by garbas** [comment]
fixed in #234
